espnet2/bin/k2_asr_inference.py

```diff
@@ -0,0 +1,467 @@
+    final_state = num_states
+    arcs = ''
+    for i in range(num_states):
+        for j in range(num_states):
+            if i == j:
+                arcs += f'{i} {i} {tokens[i]} 0 0.0\n'
+            else:
+                arcs += f'{i} {j} {tokens[j]} {tokens[j]} 0.0\n'
+        arcs += f'{i} {final_state} -1 -1 0.0\n'
+    arcs += f'{final_state}'
+    ans = k2.Fsa.from_str(arcs, num_aux_labels=1)
+    ans = k2.arc_sort(ans)
+    return ans
+
+# copied from: https://github.com/k2-fsa/snowfall/blob/master/snowfall/common.py#L309
+def get_texts(best_paths: k2.Fsa, indices: Optional[torch.Tensor] = None) -> List[List[int]]:
+    '''Extract the texts from the best-path FSAs, in the original order (before
+       the permutation given by `indices`).
+       Args:
+           best_paths:  a k2.Fsa with best_paths.arcs.num_axes() == 3, i.e.
+                    containing multiple FSAs, which is expected to be the result
+                    of k2.shortest_path (otherwise the returned values won't
+                    be meaningful).  Must have the 'aux_labels' attribute, as
+                  a ragged tensor.
+           indices: possibly a torch.Tensor giving the permutation that we used
+                    on the supervisions of this minibatch to put them in decreasing
+                    order of num-frames.  We'll apply the inverse permutation.
+                    Doesn't have to be on the same device as `best_paths`
+      Return:
+          Returns a list of lists of int, containing the label sequences we
+          decoded.
+    '''
+    # remove any 0's or -1's (there should be no 0's left but may be -1's.)
+
+    if isinstance(best_paths.aux_labels, k2.RaggedInt):
+        aux_labels = k2r.remove_values_leq(best_paths.aux_labels, 0)
+        aux_shape = k2r.compose_ragged_shapes(best_paths.arcs.shape(),
+                                                    aux_labels.shape())
+
+        # remove the states and arcs axes.
+        aux_shape = k2r.remove_axis(aux_shape, 1)
+        aux_shape = k2r.remove_axis(aux_shape, 1)
+        aux_labels = k2.RaggedInt(aux_shape, aux_labels.values())
+    else:
+        # remove axis corresponding to states.
+        aux_shape = k2r.remove_axis(best_paths.arcs.shape(), 1)
+        aux_labels = k2.RaggedInt(aux_shape, best_paths.aux_labels)
+        # remove 0's and -1's.
+        aux_labels = k2r.remove_values_leq(aux_labels, 0)
+
+    assert (aux_labels.num_axes() == 2)
+    aux_labels, _ = k2r.index(aux_labels,
+                                    invert_permutation(indices).to(dtype=torch.int32,
+                                                                   device=best_paths.device))
+    return k2r.to_list(aux_labels)
+
+
+def invert_permutation(indices: torch.Tensor) -> torch.Tensor:
+    ans = torch.zeros(indices.shape, device=indices.device, dtype=torch.long)
+    ans[indices] = torch.arange(0, indices.shape[0], device=indices.device)
+    return ans
+
+class Speech2Text:
```

Comment:
Can you create unittests like existing S2T? https://github.com/espnet/espnet/blob/e60d1786879839234cf3db8aad1f71af6f93333d/test/espnet2/bin/test_asr_inference.py#L107